File: Networking/src/main/java/project/network/protobuffprotocol/ProtoUtils.java
package project.network.protobuffprotocol;

import project.model.Flight;
import project.model.Ticket;
import project.model.User;

import java.sql.Date;
import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;

public class ProtoUtils {

    public static String getError(Protobuffs.Response response) {
        String errorMessage = response.getError();
        return errorMessage;
    }

    public static Protobuffs.Request createLoginRequest(String email, String password){
        Protobuffs.Request request= Protobuffs.Request.newBuilder().setType(Protobuffs.Request.Type.Login)
                .setEmail(email).setPassword(password).build();
        return request;
    }
    public static Protobuffs.Request createLogoutRequest(User user){
        Protobuffs.User user1=Protobuffs.User.newBuilder().setFirstName(user.getFirstName()).setLastName(user.getLastName()).setEmail(user.getEmail()).setPassword(user.getPassword()).build();
        Protobuffs.Request request= Protobuffs.Request.newBuilder().setType(Protobuffs.Request.Type.Logout)
                .setUser(user1).build();
        return request;
    }

    public static User getUser(Protobuffs.Response response) {
        User user=new User(response.getUser().getFirstName(), response.getUser().getLastName(), response.getUser().getEmail(), response.getUser().getPassword());
        return user;
    }

    public static User getUser(Protobuffs.Request request) {
        User user=new User(request.getUser().getFirstName(), request.getUser().getLastName(), request.getUser().getEmail(), request.getUser().getPassword());
        return user;
    }

    public static Integer getSeats(Integer seats) {
        return seats;
    }

    public static Protobuffs.Response createOkResponse() {
        Protobuffs.Response response=Protobuffs.Response.newBuilder()
                .setType(Protobuffs.Response.Type.Ok).build();
        return response;
    }

    public static Protobuffs.Response createErrorResponse(String message) {
        Protobuffs.Response response=Protobuffs.Response.newBuilder()
                .setType(Protobuffs.Response.Type.Error)
                .setError(message).build();
        return response;
    }

    public static Protobuffs.Response createLoginResponse(Integer id) {
        Protobuffs.Response response=  Protobuffs.Response.newBuilder()
                .setType( Protobuffs.Response.Type.Login)
                .setId(id).build();
        return response;
    }

    public static List<Flight> getFlights(Protobuffs.Response response) {
        Flight[] flights = new Flight[response.getFlightsCount()];
        for(int i=0;i<response.getFlightsCount();i++){
            Protobuffs.Flight flight=response.getFlights(i);
            LocalDateTime date = Timestamp.valueOf(flight.getDate().replace('T', ' ')).toLocalDateTime();
            Flight flight1 = new Flight(flight.getDestination(),flight.getAirport(), date,flight.getTotalSeats(), flight.getRemainingSeats());
            flight1.setId(flight.getId());
            flights[i]=flight1;
        }
        return Arrays.asList(flights.clone());
    }

    public static Protobuffs.Request createFlightRequest(String destination, String airport) {
        Protobuffs.Request request= Protobuffs.Request.newBuilder().setType(Protobuffs.Request.Type.SearchByAirport)
                .setDestination(destination).setAirport(airport).build();
        return request;
    }

    public static Protobuffs.Request createFlightRequest() {
        Protobuffs.Request request= Protobuffs.Request.newBuilder().setType(Protobuffs.Request.Type.SearchByAirport)
                .build();
        return request;
    }

    public static Protobuffs.Request createAllFlightsRequest() {
        Protobuffs.Request request= Protobuffs.Request.newBuilder().setType(Protobuffs.Request.Type.GetAllFlights)
                .build();
        return request;
    }

    public static Protobuffs.Response createFlightResponse(Iterable<Flight> flights) {
        Protobuffs.Response.Builder response= Protobuffs.Response.newBuilder()
                .setType(Protobuffs.Response.Type.SearchByAirport);
        for (Flight flight: flights){
            Protobuffs.Flight flight1 = Protobuffs.Flight.newBuilder().setId(flight.getId()).setDestination(flight.getDestination())
                    .setAirport(flight.getAirport()).setDate(flight.getDateTime().toString()).setTotalSeats(flight.getTotalSeats()).setRemainingSeats(flight.getRemainingSeats()).build();
            response.addFlights(flight1);
        }

        return response.build();
    }

    public static Protobuffs.Response createAllFlightsResponse(Iterable<Flight> flights) {
        Protobuffs.Response.Builder response= Protobuffs.Response.newBuilder()
                .setType(Protobuffs.Response.Type.GetAllFlights);
        for (Flight flight: flights){
            Protobuffs.Flight flight1 = Protobuffs.Flight.newBuilder().setId(flight.getId()).setDestination(flight.getDestination())
                    .setAirport(flight.getAirport()).setDate(flight.getDateTime().toString()).setTotalSeats(flight.getTotalSeats()).setRemainingSeats(flight.getRemainingSeats()).build();
            response.addFlights(flight1);
        }

        return response.build();
    }

    public static Protobuffs.Request createSellTicketRequest(Ticket ticket) {
        Protobuffs.Ticket ticket1 = Protobuffs.Ticket.newBuilder().setFlightId(ticket.getFlightId())
                .setName(ticket.getName()).setTourists(ticket.getTourists()).setAddress(ticket.getAddress()).setSeats(ticket.getSeats()).build();
        Protobuffs.Request request= Protobuffs.Request.newBuilder().setType(Protobuffs.Request.Type.SellTicket)
                .setTicket(ticket1).build();
        return request;
    }

    public static Ticket getTicket(Protobuffs.Ticket ticket) {
        Ticket ticket1 = new Ticket(ticket.getFlightId(), ticket.getName(), ticket.getTourists(), ticket.getAddress(), ticket.getSeats());
        return  ticket1;
    }

    public static Flight getFlight(Protobuffs.Flight flight){
        LocalDateTime date = Timestamp.valueOf(flight.getDate().replace('T', ' ')).toLocalDateTime();
        Flight flight1 = new Flight(flight.getDestination(),flight.getAirport(), date,flight.getTotalSeats(), flight.getRemainingSeats());
        flight1.setId(flight.getId());
        return flight1;
    }

    public static Protobuffs.Response createTicketSoldResponse(List<Flight> flights) {
        Protobuffs.Response.Builder response= Protobuffs.Response.newBuilder()
                .setType(Protobuffs.Response.Type.SellTicket);
        for (Flight flight: flights){
            Protobuffs.Flight flight1 = Protobuffs.Flight.newBuilder().setId(flight.getId()).setDestination(flight.getDestination())
                    .setAirport(flight.getAirport()).setDate(flight.getDateTime().toString()).setTotalSeats(flight.getTotalSeats()).setRemainingSeats(flight.getRemainingSeats()).build();
            response.addFlights(flight1);
        }
        return response.build();
    }

}

File: Protobuf/ProtoUtils.cs
using System;
using System.Collections.Generic;
using System.Text;
using Flight = Model.Flight;
using Ticket = Model.Ticket;
using User = Model.User;
using proto = Project.Network.Protobuffprotocol;
using Project.Network.Protobuffprotocol;

namespace Protobuf
{
    public class ProtoUtils
    {
        public static String getError(proto.Response response)
        {
            String errorMessage = response.Error;
            return errorMessage;
        }

        public static Request createLoginRequest(String email, String password)
        {
            Request request = new proto.Request { Type = proto.Request.Types.Type.Login, Email = email, Password = password };
            return request;
        }

        public static proto.Request createLogoutRequest(User user)
        {
            proto.User acc = new proto.User { FirstName = user.FirstName, LastName = user.LastName, Email = user.Email,  Password = user.Password };
            Request request = new proto.Request { Type = proto.Request.Types.Type.Logout, User = acc };
            return request;
        }

        public static User getUser(proto.Response response)
        {
            User user = new User(response.User.FirstName, response.User.LastName, response.User.Email, response.User.Password);
            //user.Id = response.User.Id;
            return user;
        }

        public static User getUser(proto.Request request)
        {
            User user = new User(request.User.FirstName, request.User.LastName, request.User.Email, request.User.Password);
            return user;
        }

        public static proto.Response createOkResponse()
        {
            proto.Response response = new proto.Response { Type = proto.Response.Types.Type.Ok };
            return response;
        }

        public static proto.Response createErrorResponse(String message)
        {
            proto.Response response = new proto.Response { Type = proto.Response.Types.Type.Error };
            return response;
        }

        public static proto.Response createLoginResponse(int id)
        {
            proto.Response response = new proto.Response { Type = proto.Response.Types.Type.Login, Id = id };

            return response;
        }

        public static IList<Flight> getFlights(proto.Response response)
        {
            Flight[] flights = new Flight[response.Flights.Count];
            for (int i = 0; i < response.Flights.Count; i++)
            {
                proto.Flight flight = response.Flights[i];
                Flight flight1 = new Flight(flight.Destination, flight.Airport, DateTime.Parse(flight.Date), (int) flight.TotalSeats,(int) flight.RemainingSeats);
                flight1.Id = flight.Id;
                flights[i] = flight1;
            }

            return flights;
        }

        public static proto.Request createFlightRequest(String destination, String airport)
        {
            proto.Request request = new proto.Request { Type = proto.Request.Types.Type.SearchByAirport, Destination = destination, Airport = airport };
            return request;
        }
        public static proto.Request createFlightRequest()
        {
            proto.Request request = new proto.Request { Type = proto.Request.Types.Type.SearchByAirport };
            return request;
        }

        public static proto.Request createAllFlightsRequest()
        {
            proto.Request request = new proto.Request { Type = proto.Request.Types.Type.GetAllFlights };
            return request;
        }

        public static proto.Response createFlightResponse(IList<Flight> flights)
        {
            proto.Response response = new proto.Response { Type = proto.Response.Types.Type.SearchByAirport };
            foreach (var flight in flights)
            {
                proto.Flight flight1 = new proto.Flight
                {            
                    Destination = flight.Destination,
                    Airport = flight.Airport,
                    Date = flight.Date_Time.ToString("yyyy-MM-dd hh:mm:ss"),
                    TotalSeats = flight.TotalSeats,
                    RemainingSeats = flight.RemainingSeats
                };
                flight1.Id = flight.Id;
                response.Flights.Add(flight1);
            }

            return response;
        }

        public static proto.Response createAllFlightsResponse(IList<Flight> flights)
        {
            proto.Response response = new proto.Response { Type = proto.Response.Types.Type.GetAllFlights };
            foreach (var flight in flights)
            {
                proto.Flight flight1 = new proto.Flight
                {
                    Destination = flight.Destination,
                    Airport = flight.Airport,
                    Date = flight.Date_Time.ToString("yyyy-MM-dd hh:mm:ss"),
                    TotalSeats = flight.TotalSeats,
                    RemainingSeats = flight.RemainingSeats
                };
                flight1.Id = flight.Id;
                response.Flights.Add(flight1);
            }

            return response;
        }

        public static proto.Request createSellTicketRequest(Ticket ticket)
        {
            proto.Ticket ticket1 = new proto.Ticket
            { FlightId = ticket.FlightId, Name = ticket.Name, Tourists = ticket.Tourists, Address = ticket.Address, Seats = ticket.Seats };
            proto.Request request = new Request { Type = Request.Types.Type.SellTicket, Ticket = ticket1 };
            return request;
        }

        public static Ticket getTicket(proto.Ticket ticket)
        {
            Ticket ticket1 = new Ticket(ticket.FlightId, ticket.Name, ticket.Tourists, ticket.Address, ticket.Seats);
            return ticket1;
        }

        public static proto.Response createTicketSoldResponse(IList<Flight> flights)
        {
            proto.Response response = new proto.Response { Type = proto.Response.Types.Type.SellTicket };
            foreach (var flight in flights)
            {
                proto.Flight flight1 = new proto.Flight
                {
                    Destination = flight.Destination,
                    Airport = flight.Airport,
                    Date = flight.Date_Time.ToString("yyyy-MM-dd hh:mm:ss"),
                    TotalSeats = flight.TotalSeats,
                    RemainingSeats = flight.RemainingSeats
                };
                flight1.Id = flight.Id;
                response.Flights.Add(flight1);
            }

            return response;
        }
    }
}